README.md
# menu_button

[![Pub Package](https://img.shields.io/pub/v/menu_button.svg?style=for-the-badge&color=blue)](https://pub.dartlang.org/packages/menu_button)
![License](https://img.shields.io/badge/license-MIT-blue.svg?style=for-the-badge&color=blue)

Flutter widget to display a popup menu button very simply and easily customizable.

## Resources

- [Documentation](https://pub.dev/documentation/menu_button/latest/menu_button/MenuButton-class.html)
- [Pub Package](https://pub.dev/packages/menu_button)
- [GitHub Repository](https://github.com/huextrat/menu_button)

## Installations

Add `menu_button: ^1.4.2+1` in your `pubspec.yaml` dependencies. And import it:

```dart
import 'package:menu_button/menu_button.dart';
```

## Usage

The widget has a lot of properties to customize it, we will see here the ones needed to get a "basic" menu button.

Firstly we have to declare a variable to keep the selected item (`selectedKey`) and a list that contains all the items we want to display in this menu button.

Here we will make a list of strings that we will call `keys` and that contains the values `Low`, `Medium` & `High`.

```dart
String selectedKey;

List<String> keys = <String>[
  'Low',
  'Medium',
  'High',
];
```

Now that we have these two elements we can start using the `MenuButton<T>` widget.

```dart
MenuButton<String>(
  child: normalChildButton,
  items: keys,
  itemBuilder: (String value) => Container(
   height: 40,
   alignment: Alignment.centerLeft,
   padding: const EdgeInsets.symmetric(vertical: 0.0, horizontal: 16),
   child: Text(value),
  ),
  toggledChild: Container(
    child: normalChildButton,
  ),
  onItemSelected: (String value) {
    setState(() {
      selectedKey = value;
    });
  },
  onMenuButtonToggle: (bool isToggle) {
    print(isToggle);
  },
)
```

And finally here is an example of the `child` widget used for the MenuButton above:

```dart
final Widget normalChildButton = SizedBox(
  width: 93,
  height: 40,
  child: Padding(
    padding: const EdgeInsets.only(left: 16, right: 11),
    child: Row(
      mainAxisAlignment: MainAxisAlignment.spaceBetween,
      children: <Widget>[
        Flexible(
          child: Text(selectedKey, overflow: TextOverflow.ellipsis)
        ),
        const SizedBox(
          width: 12,
          height: 17,
          child: FittedBox(
            fit: BoxFit.fill,
            child: Icon(
              Icons.arrow_drop_down,
              color: Colors.grey,
            ),
          ),
        ),
      ],
    ),
  ),
);
```

Of course you can make your own according to your needs.

## Basic Parameters

| Parameter | Description |
|---|---|
| `child` | A widget to display the default button to trigger the menu button |
| `items` | The list that contains all values that you want to display on the menu button |
| `itemBuilder` | A widget to design each item of the menu button |
| `onItemSelected` | Function triggered when an item is selected |
| `onMenuButtonToggle` | Function triggered when menu button is triggered (`true` if displayed, `false` if not) |
| `toggledChild` | Same as child but when the menu button is opened |

## More Parameters

| Parameter | Description |
|---|---|
| `crossTheEdge` | By default `false` you can set it to `true` if you want the button to expand |
| `divider` | A custom divider between each items |
| `decoration` | A custom decoration for menu button |
| `edgeMargin` | By default `0` add a custom value to prevent the button to not touch the edge, check the example `edge_menu_button.dart` for more information |
| `itemBackgroundColor` | By default `Colors.white` add custom Colors to customize the background of every items |
| `label` | Add a widget to display a custom label as MaterialDesign on top of the button, check `label_menu_button.dart` for more information |
| `labelDecoration` | If you use a `label` you can set a custom `LabelDecoration` |
| `menuButtonBackgroundColor` | By default `Colors.white` add custom Colors to customize the background of the menu button |
| `popupHeight` | By default `popupHeight` is automatically calculated but if you need a custom height use this property |
| `scrollPhysics` | By default items are not scrollable (`NeverScrollableScrollPhysics`), add a ScrollPhysics to enable it, for instance `AlwaysScrollableScrollPhysics` |
| `showSelectedItemOnList` | By default `true`, set it to `false` if you don't want the selected items in the list |

---

For a more detail example please take a look at the `example` folder.

## Example

Menu button with 3 items:

<img src="https://raw.githubusercontent.com/huextrat/menu_button/master/example/new_example.gif" width="400" height="790">

---

If something is missing, feel free to open a ticket or contribute!


CHANGELOG.md
## 1.4.2+2 - 2022-03-23

- Added `DropVerticalDirection` in `MenuButton`

## 1.4.2+1 - 2021-03-23

- Fix InkWell when radius is applied to decoration
- `onMenuButtonToggle` is no more required

## 1.4.1 - 2021-03-09

- More precise API Docs

## 1.4.0+1 - 2021-03-03

- Null safety
- Improve documentation
- Tested on Flutter 2.0 it's working on every platform

## 1.3.1+1 - 2021-03-03

- Update README to make it clearer

## 1.3.1 - 2021-02-18

- Fix: add the item background color by default when the list is scrollable and no color is defined in `decoration` property
- Fix: `showSelectedItemOnList` is now working as expected
- Add animation when `crossTheEdge` is true

## 1.3.0 - 2021-02-17

- Add `itemBackgroundColor` & `menuButtonBackgroundColor` to enable ripple effect when you press an item
It's highly recommended to use them instead of the decoration color.

- `showSelectedItemOnList` property instead of `dontShowTheSameItemSelected` and the default value is now `true`

## 1.2.1 - 2020-06-16

Fix width when MenuButton is on Expanded widget

## 1.2.0+1 - 2020-06-16

MenuButton now display the decoration color everywhere

## 1.2.0 - 2020-06-16

Breaking change
- Default value of `dontShowTheSameItemSelected` is now `false`
If you are using it in previous version with its default value please do not forget to add the property to `true`

## 1.1.1 - 2020-05-19
Adjust offset depending on verticalMenuPadding

## 1.1.0 - 2020-05-18
Add menu button without showing the current item
Add a label property if needed

## 1.0.0 - 2020-05-08
Add crossEdge & edgeMargin configuration
Update example

## 0.2.1 - 2020-04-14 
Define scrollPhysics in MenuButton is now available

## 0.1.1+1 - 2020-03-25

Add `onMenuButtonToggle` callback to find out if menu button is open or not

## 0.1.0+1 - 2020-01-20

Fix README.md

## 0.1.0 - 2020-01-20

First version of `menu_button` available 

## 0.0.2 - 2020-01-20

Initial version

example/lib/usages/label-menu-button.dart
import 'package:flutter/material.dart';
import 'package:menu_button/menu_button.dart';

class MenuButtonLabel extends StatefulWidget {
  const MenuButtonLabel({
    Key key,
    @required this.theme,
  }) : super(key: key);

  final ThemeData theme;

  @override
  _MenuButtonLabelState createState() => _MenuButtonLabelState();
}

class _MenuButtonLabelState extends State<MenuButtonLabel> {
  String selectedKey;
  String initialValue;

  List<String> keys = [
    'Low',
    'Medium',
    'High',
  ];

  @override
  void initState() {
    initialValue = keys[0];
    selectedKey = keys[0];
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    final Widget childButtonWithoutSameItem = Container(
      padding: const EdgeInsets.only(left: 16, right: 11),
      width: 93,
      height: 40,
      child: Row(
        mainAxisAlignment: MainAxisAlignment.spaceBetween,
        children: <Widget>[
          Flexible(
            child: Text(
              selectedKey,
              overflow: TextOverflow.ellipsis,
            ),
          ),
          SizedBox(
            width: 12,
            height: 17,
            child: FittedBox(
              fit: BoxFit.fill,
              child: Icon(
                Icons.arrow_drop_down,
                color: Colors.grey,
              ),
            ),
          ),
        ],
      ),
    );

    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        Padding(
          padding: const EdgeInsets.only(bottom: 12.0),
          child: Text(
            'Menu button usage with label',
            style: widget.theme.textTheme.headline2.copyWith(
              fontSize: 18,
            ),
          ),
        ),
        MenuButton(
          child: childButtonWithoutSameItem,
          items: keys,
          topDivider: true,
          dontShowTheSameItemSelected: true,
          selectedItem: selectedKey,
          label: Text(
            'Label',
            style: TextStyle(
              fontSize: 12,
            ),
          ),
          labelDecoration: LabelDecoration(
            verticalMenuPadding: 12,
            // background: Colors.red
          ),
          itemBuilder: (value) => Container(
            height: 40,
            alignment: Alignment.centerLeft,
            padding: const EdgeInsets.symmetric(vertical: 0.0, horizontal: 16),
            child: Text(value),
          ),
          toggledChild: Container(
            child: childButtonWithoutSameItem,
          ),
          divider: Container(
            height: 1,
            color: Colors.grey,
          ),
          onItemSelected: (value) {
            setState(() {
              selectedKey = value;
            });
          },
          decoration: BoxDecoration(
            border: Border.all(color: Colors.grey[300]),
            borderRadius: const BorderRadius.all(
              Radius.circular(3.0),
            ),
          ),
          itemBackgroundColor: Colors.blue,
          menuButtonBackgroundColor: Colors.green,
          onMenuButtonToggle: (isToggle) {
            print(isToggle);
          },
        ),
      ],
    );
  }
}
